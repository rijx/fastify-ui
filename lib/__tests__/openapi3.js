const { validate } = require("openapi-schema-validation");
const yaml = require("js-yaml");

const { generateConfig } = require("../openapi3");

function testRoute(route, useBaseConfig = true) {
  const config = generateConfig({
    routes: [route].filter(Boolean),
    config: useBaseConfig
      ? {
          info: {
            title: "Test",
            version: "0.0.0"
          }
        }
      : null
  });

  validate(config, 3);

  expect(yaml.safeDump(config, { skipInvalid: true })).toMatchSnapshot();
}

describe("Fastify UI", () => {
  describe("Open API 3 generator", () => {
    test("empty config", () => {
      testRoute(null, false);
    });

    test("simple route without schema", () => {
      testRoute({
        method: "POST",
        url: "/test/:random",
        handler: jest.fn()
      });
    });

    test("simple route with a schema just to hide it", () => {
      testRoute({
        method: "POST",
        url: "/test/:random",
        schema: {
          hide: true
        },
        handler: jest.fn()
      });
    });

    test("simple route using the shorthand syntax", () => {
      testRoute({
        method: "POST",
        url: "/test/:random",
        schema: {
          params: {
            random: { type: "string" }
          }
        },
        handler: jest.fn()
      });
    });

    test("advanced route", () => {
      testRoute({
        method: "POST",
        url: "/test/:random",
        schema: {
          operationId: "testMethod",
          summary: "Tests the basic features",
          params: {
            required: ["random"],
            type: "object",
            properties: {
              random: { type: "string" }
            }
          },
          body: {
            type: "object",
            properties: {
              name: { type: "string" },
              profession: { type: "string" }
            }
          },
          response: {
            200: {
              description: "The test ran successfully.",
              type: "object",
              properties: {
                success: {
                  type: "boolean"
                }
              }
            },
            500: {
              description: "An error occurred."
            }
          },
          security: [
            {
              bearerAuth: ["tests:run"]
            }
          ]
        },
        handler: jest.fn()
      });
    });
  });
});
