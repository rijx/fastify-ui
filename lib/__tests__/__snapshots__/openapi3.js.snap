// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fastify UI Open API 3 generator advanced route 1`] = `
"openapi: 3.0.0
info:
  title: Test
  version: 0.0.0
paths:
  '/test/{random}':
    post:
      operationId: testMethod
      summary: Tests the basic features
      security:
        - bearerAuth:
            - 'tests:run'
      requestBody:
        content:
          application/json:
            schema:
              type: object
              properties:
                name:
                  type: string
                profession:
                  type: string
      responses:
        '200':
          description: The test ran successfully.
          content:
            application/json:
              schema:
                description: The test ran successfully.
                type: object
                properties:
                  success:
                    type: boolean
        '500':
          description: An error occurred.
      parameters:
        - name: random
          schema:
            type: string
          required: true
          in: path
"
`;

exports[`Fastify UI Open API 3 generator empty config 1`] = `
"openapi: 3.0.0
paths: {}
"
`;

exports[`Fastify UI Open API 3 generator simple route using the shorthand syntax 1`] = `
"openapi: 3.0.0
info:
  title: Test
  version: 0.0.0
paths:
  '/test/{random}':
    get:
      responses: {}
      parameters:
        - name: random
          schema:
            type: string
          in: path
"
`;

exports[`Fastify UI Open API 3 generator simple route with a schema just to hide it 1`] = `
"openapi: 3.0.0
info:
  title: Test
  version: 0.0.0
paths: {}
"
`;

exports[`Fastify UI Open API 3 generator simple route without schema 1`] = `
"openapi: 3.0.0
info:
  title: Test
  version: 0.0.0
paths:
  '/test/{random}':
    post:
      summary: ''
"
`;
